cmd/warpforge/watch.go

```diff
@@ -23,20 +32,132 @@ var watchCmdDef = cli.Command{
 	Action: util.ChainCmdMiddleware(cmdWatch,
 		util.CmdMiddlewareLogging,
 		util.CmdMiddlewareTracingConfig,
+		util.CmdMiddlewareCancelOnInterrupt,
 	),
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "socket",
+			Usage: "Experimental flag for getting execution status of the watched plot externally via unix socket",
+		},
+	},
 }
 
-func cmdWatch(c *cli.Context) error {
-	if c.Args().Len() != 1 {
-		return fmt.Errorf("invalid args")
+type watchCfg struct {
+	path         string
+	recursive    bool
+	socket       bool
+	plotCfg      wfapi.PlotExecConfig
+	recentStatus int
+}
+
+func (c *watchCfg) handle(ctx context.Context, conn net.Conn) error {
+	// in lieu of doing anything complicated, shoving a status int down the pipe is sufficient
+	// for the unix socket implementation we can now use netcat or similar for a status.
+	// I.E. nc -U ./sock
+	// OR socat - UNIX-CONNECT:./sock
+	defer conn.Close()
+	enc := json.NewEncoder(conn)
+	err := enc.Encode(c.recentStatus)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *watchCfg) serveLoop(ctx context.Context, l net.Listener) error {
+	log := logging.Ctx(ctx)
+	for {
+		conn, err := l.Accept() // blocks, doesn't accept a context.
+		if err != nil {
+			return err
+		}
+		go func() {
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			defer func() {
+				r := recover()
+				if r != nil {
+					log.Info("", "socket handler panic: %s", r)
+				}
+			}()
+			if err := c.handle(ctx, conn); err != nil {
+				log.Info("", "socket handler: %s", err.Error())
+			}
+		}()
+		select {
+		case <-ctx.Done():
+			log.Info("", "socket no longer accepting connections")
+			return nil
+		default:
+		}
+	}
+}
+
+//serve will create and listen to a unix socket on the given socket path
+func (c *watchCfg) serve(ctx context.Context, sockPath string) error {
+	log := logging.Ctx(ctx)
+	cfg := net.ListenConfig{}
+	l, err := cfg.Listen(ctx, "unix", sockPath)
+	if err != nil {
+		return err
+	}
+	// unix socks: they have permissions systems.
+	os.Chmod(sockPath, 0777) // ignore err?
+
+	result := make(chan error)
+	go func() {
+		result <- c.serveLoop(ctx, l)
+	}()
+	<-ctx.Done()
+	l.Close()
+	log.Info("", "listener closed")
+	return <-result
+}
+
+func isSocket(m fs.FileMode) bool {
+	return m&fs.ModeSocket != 0
+}
+
+// rmUnixSocket will perform a racy, lockless "liveness" check on the socket before removing it.
+func (c *watchCfg) rmUnixSocket(path string) error {
```

Comment:
Needs more description of expected outcomes.

- If a socket is found, but someone's alive and answering on it already -> what happens?
- If a socket is found, but someone's _not_ answering it (explicit error) -> what happens?
- If a socket is found, but someone's not answering it in a timely fashion -> what happens?

And, from the other side, because this is a coordination problem scene:

- If another process _is_ alive, but failed to answer in a timely fashion, and we removed its socket -> what happens to it?  Does it notice?  What does it do when it does?

Reply:
We talked about file locking in chat earlier, so I'm curious -- is this the function you were thinking of wrapping it around?

Would it be able to make things significantly better?

(I think we may end up being okay merging something that's a first pass on this and has some admitted race conditions, as long we have it reasonably encapsulated so we can revisit it later.  Part of the reason I'm okay with this is because I actually don't know how solvable the TOCTOUs here _are_, in the limit.  But if you did end up with more notes on where exactly the future work could go, they might be worth attaching.)

Reply:
Yes, the file locking was being used for the same purpose as this function. Would file locks actually make it _better_? Probably not in any significant way. I think the underlying coordination problem would still exist but just present in a slightly different way and possibly require user intervention to unlock in edge cases. So no; that's worse.